src/contrib/talloc.rst

```diff
@@ -0,0 +1,394 @@
+Memory allocator: talloc
+########################
+
+`Talloc`_ is a hierarchical memory allocator which is used heavily inside SSSD
+instead of ``malloc``. It allows you to manage memory of complex data structures
+more naturally *by defining an oriented graph of memory context* and its
+descendants and thus simplifying the data flow and error management. It also
+extends the allocator with runtime type safety and destructors.
+
+.. seealso::
+
+    The talloc library provides many useful features and mastering it will help
+    you to simplify the code and its logic a lot. However, this document
+    provides only a basic description of talloc and its use cases and patterns
+    inside the SSSD project. It is recommended for you to :tag:`strong` read the
+    official `Talloc Tutorial`_ to learn more about its features.
+    :end-tag:`strong`
+
+.. _Talloc: https://talloc.samba.org
+.. _Talloc Tutorial: https://talloc.samba.org/talloc/doc/html/libtalloc__tutorial.html
+
+Memory hierarchy
+****************
+
+Talloc organizes allocated memory into an oriented graph. This makes it very
+easy to free complex data structures. See the following example to get the idea:
+
+.. code-tabs::
+
+    .. code-tab:: c
+        :label: talloc
+
+        struct record {
+            char *name;
+            char *value;
+        };
+
+        struct record *r;
+
+        r = talloc_zero(NULL, struct record);
+        r->name = talloc_strdup(r, "Hello");
+        r->value = talloc_strdup(r, "World");
+
+        talloc_free(r);
+
+    .. code-tab:: c
+        :label: malloc
+
+        struct record {
+            char *name;
+            char *value;
+        };
+
+        struct record *r;
+
+        r = talloc_zero(NULL, struct record);
+        r->name = talloc_strdup(r, "Hello");
+        r->value = talloc_strdup(r, "World");
+
+        free(r->name);
+        free(r->value);
+        free(r);
+
+We created a simple hierarchy of memory contexts by associating ``name`` and
+``value`` with the record ``r``. This allowed us to free the whole structure
```

Comment:
This can use the present tense, instead of past tense.

s/created/create/
s/allowed/allows/